src/PixelGraph.cpp

```diff
@@ -509,4 +509,337 @@ int PixelGraph::calculateValence(cv::Point p) {
     if (n->bottomRight != NULL) connections++;
 
     return connections;
+}
+
+void PixelGraph::calculateFillType(std::vector<std::vector<FillType>>& fills) {
+    // iterate over the graph. We will be considering 3x3 sections
+    for (int y = 0; y < graph->size(); y++) {
+        for (int x = 0; x < graph->at(0)->size(); x++) {
+            fills[y][x] = FillType::NONE;
+
+            Node* n = graph->at(y)->at(x);
+            int valence = calculateValence(cv::Point(x, y));
+
+            if (valence == 0) {
+                fills[y][x] = FillType::FULL;
+            } else if (valence == 1) {
+                // either diagonal end or cardinal end
+                if (n->top != NULL || n->left != NULL || n->right != NULL || n->bottom != NULL) {
+                    fills[y][x] = FillType::CARD_END;
+                }
+                else {
+                    fills[y][x] = FillType::DIAG_END;
+                }
+            }
+            else if (valence == 2) {
+                if ((n->top != NULL && n->bottom != NULL) || (n->left != NULL && n->right != NULL)) {
+                    // straight line through cardinals
+                    fills[y][x] = FillType::FULL;
+                }
+                else if ((n->topLeft != NULL && n->bottomRight != NULL) || (n->topRight != NULL && n->bottomLeft != NULL)) {
+                    // true diagonal
+                    fills[y][x] = FillType::DIAGONAL;
+                }
+                else {
+                    // now we need to check for diagonals!
+                    Node* top = (y - 1.0 < 0) ? NULL : graph->at(y - 1.0)->at(x);
+                    Node* bottom = (y + 1.0 >= graph->size()) ? NULL : graph->at(y + 1.0)->at(x);
+
+                    bool nHasDiagonals = n->bottomLeft != NULL || n->bottomRight != NULL || n->topLeft != NULL || n->topRight != NULL;
+                    bool topDiagonals = top != NULL && (top->bottomLeft != NULL || top->bottomRight != NULL);
+                    bool bottomDiagonals = bottom != NULL && (bottom->topLeft != NULL || bottom->topRight != NULL);
+                    bool otherDiagonals = topDiagonals || bottomDiagonals;
+                    
+                    if (nHasDiagonals || otherDiagonals) {
+                        fills[y][x] = FillType::NO_CORNER;

```

Comment:
Wait, so what if this is an elbow? Should there be a case for missing two corners?

x o x
o x o
o o o

Reply:
Responding to the general question here as well! I had to infer a lot of the `FillType` conditions myself, but for the most part, that just meant cross referencing the sample images in the paper and testing some things out on paper to make sure that they worked well enough. The report authors essentially got to the point of "we made a voronoi, you can skip actually making the real voronoi, just use a quick scanline algorithm to do it instead". They also say that there are a fairly limited number of shapes that you can end up with, which is definitely the case. In a lot of scenarios it's just a matter of turning things appropriately.

As far as elbows -- that's a good point, missing 2 corners isn't really handled here. In this scenario, we're checking for the scenario of a full diagonal or a diagonal within the neighbourhood that isn't attached to the pixel in question, so the cases for
```
x - -         - x -
- x -         a x x
- - x         - a -
```
I'll add the elbow case now.

Reply:
What about this?

- x x
a x a
- a -

In this case, n has a cardinal and a diagonal, yet two corners are missing